CoreAardvark/ARKExceptionLogging.m

```diff
@@ -0,0 +1,78 @@
+//
+//  ARKExceptionLogging.m
+//  Aardvark
+//
+//  Created by Nick Entin on 9/7/17.
+//  Copyright 2017 Square, Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+#import "ARKExceptionLogging.h"
+
+#import "ARKLogDistributor.h"
+#import "ARKLogDistributor_Protected.h"
+#import "ARKLogging.h"
+
+
+NSUncaughtExceptionHandler *_Nullable ARKPreviousUncaughtExceptionHandler = nil;
+
+NSMutableArray *_Nullable ARKUncaughtExceptionLogDistributors = nil;
+
+void ARKHandleUncaughtException(NSException *exception)
+{
+    for (ARKLogDistributor *logDistributor in ARKUncaughtExceptionLogDistributors) {
+        [logDistributor logWithType:ARKLogTypeError userInfo:nil format:@"Uncaught exception '%@':\n%@", exception.name, exception.debugDescription];
+        [logDistributor distributeAllPendingLogsWithCompletionHandler:^{}];
+        [logDistributor waitUntilAllPendingLogsHaveBeenDistributed];
+    }
+    
+    if (ARKPreviousUncaughtExceptionHandler != nil) {
+        ARKPreviousUncaughtExceptionHandler(exception);
+    }
+}
+
+void ARKEnableLogOnUncaughtException()
+{
+    ARKEnableLogOnUncaughtExceptionToLogDistributor([ARKLogDistributor defaultDistributor]);
+}
+
+void ARKEnableLogOnUncaughtExceptionToLogDistributor(ARKLogDistributor *_Nonnull logDistributor)
+{
+    if (ARKUncaughtExceptionLogDistributors == nil) {
+        ARKUncaughtExceptionLogDistributors = [NSMutableArray arrayWithObject:logDistributor];
+
+        ARKPreviousUncaughtExceptionHandler = NSGetUncaughtExceptionHandler();
```

Comment:
I think you need a stack of `ARKPreviousUncaughtExceptionHandler`s. Otherwise the following code
```
ARKEnableLogOnUncaughtException()
ARKEnableLogOnUncaughtExceptionToLogDistributor(mySpecialDistributor)
```
will nuke whatever uncaught exception handler was previously set.

Reply:
Also, if you do the above, how many times will `ARKHandleUncaughtException` (which logs to all registered log distributors) be called? I think twice?

Reply:
It should only ever get called once.  Unless I missed something (or they get called on different threads, as you pointed out above), it should be in one of two states:

1.  Exception logging is disabled.  `ARKHandleUncaughtException` is _not_ the specified uncaught exception handler.  `ARKUncaughtExceptionLogDistributors` is `nil`.  `ARKPreviousUncaughtExceptionHandler` is `nil`.

2.  Exception logging is enabled.  `ARKHandleUncaughtException` is the specified uncaught exception handler.  `ARKUncaughtExceptionLogDistributors` has at least one log distributor.  `ARKPreviousUncaughtExceptionHandler` is set to the previously-set uncaught exception handler.

When you enable logging and there are no existing log distributors set up, it will perform the initial setup (setting itself as the uncaught exception handler, and storing the existing one to chain).  When you enable logging on additional distributors, it will only add them to the list of distributors to use.

When you disable logging, it will remove the log distributor from the list.  Then, it checks if the list is now empty, and if so reverses the setup process, removing itself as the uncaught exception handler (and restoring the previous one) and clearing out the distributor list.

FWIW, I tested calling enable twice, and it logged the message twice, then called the existing handler.

Reply:
🤦‍♂️ I missed the `ARKUncaughtExceptionLogDistributors == nil` check somehow. Thank you for the detailed explanation. I agree that this works and my fears are unfounded. This is a really simple/clever way to make this work.

A comment above this line saying "we only want to register ourselves as an exception handler the first time a log distributor is registered" might help with long-term maintenance, but so would reading more carefully 🙂